### ISSUE TITLE
KafkaListener SpEL support for array of topics

### ISSUE THREAD

**Issue by umekong** [spring-projects/spring-kafka#221]
If I specify an array of topics within the `@KafkaListener` annotation, then I receive all messages sent to topics specified within the array:
```java
@KafkaListener(topics = { "inbound_topic_a", "inbound_topic_b" })
public void doSomething(final String message) {
    // do something with the message
}
```

However, when I modify the parameter to evaluate the value of a property using SpEL, as below:
```java
@KafkaListener(topics = "${kafka.inbound.topics}")
public void doSomething(final String message) {
    // do something with the message
}
```
and I specify a comma-separated list of topics as the value of the `kafka.inbound.topics` property; `System.setProperty("kafka.inbound.topics", "inbound_topic_a,inbound_topic_b")`, I failed to receive any messages sent to either topic.
I do however see a WARN level message:
```console
15:38:59.418 [org.springframework.kafka.KafkaListenerEndpointContainer#0-0-kafka-consumer-1] WARN  o.apache.kafka.clients.NetworkClient - Error while fetching metadata with correlation id 1 : {inbound_topic_a,inbound_topic_b=INVALID_TOPIC_EXCEPTION}
```
After much debugging, I confirmed that the SpEL expression specified within the `topics` parameter was being evaluated by the `KafkaListenerAnnotationBeanPostProcessor` as an array containing the comma-separated string as its only value.

I thought I could get around this by modifying the SpEL to split the string based on a comma character delimiter, after all the retun type is an array and that's what I'm after:
```java
@KafkaListener(topics = "#{'${kafka.inbound.topics}'.split(',')}")
public void doSomething(final String message) {
    // do something
}
```
However, this causes an `IllegalArgumentException` to be thrown:
```console
Caused by: java.lang.IllegalArgumentException: @KafKaListener can't resolve '[Ljava.lang.String;@1cbc5693' as a String
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.resolveAsString(KafkaListenerAnnotationBeanPostProcessor.java:549)
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.resolveTopics(KafkaListenerAnnotationBeanPostProcessor.java:433)
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.processListener(KafkaListenerAnnotationBeanPostProcessor.java:380)
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.processKafkaListener(KafkaListenerAnnotationBeanPostProcessor.java:339)
	at org.springframework.kafka.annotation.KafkaListenerAnnotationBeanPostProcessor.postProcessAfterInitialization(KafkaListenerAnnotationBeanPostProcessor.java:273)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsAfterInitialization(AbstractAutowireCapableBeanFactory.java:422)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1588)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:553)
	... 41 common frames omitted
```

Looking at the code for the `KafkaListenerAnnotationBeanPostProcessor`, I'm sure that because the type check on the `resolvedValue` being a `String` is an independent `if` statement (rather than an `else if`), the code block for the `resolveAsString` method will always end up throwing an `IllegalArgumentException` even though the SpEL expression that generates an array works as expected.

**Comment by artembilan** [comment]
Confirmed.
We really should have `else if` there:
```
if (resolvedValue instanceof String[]) {
	for (Object object : (String[]) resolvedValue) {
		resolveAsString(object, result);
	}
}
if (resolvedValue instanceof String) {
	result.add((String) resolvedValue);
}
```
So, now after resolving it as an array we go to other conditions rather than `return`. 😢

**Comment by garyrussell** [comment]
Strange; this code was based on the [rabbit version](https://github.com/spring-projects/spring-amqp/blob/master/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java#L469).

And there is a test for that case [here](https://github.com/spring-projects/spring-amqp/blob/master/spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/EnableRabbitIntegrationTests.java#L699) so I don't know how this would have evolved differently.

**Comment by artembilan** [comment]
I will take a look into Spring AMQP a bit later.

**Comment by garyrussell** [comment]
Looks like a day 1 bug (mea culpa).

We should probably replicate all the rabbit test cases (that apply) in spring-kafka; if I had done that, we would have caught this bug.

**Comment by artembilan** [comment]
OK.
Look what we do in the Spring AMQP:
```
Object resolvedValueToUse = resolvedValue;
if (resolvedValue instanceof String[]) {
	resolvedValueToUse = Arrays.asList((String[]) resolvedValue);
}
if (resolvedValueToUse instanceof Queue) {
	result.add(((Queue) resolvedValueToUse).getName());
}
else if (resolvedValueToUse instanceof String) {
	result.add((String) resolvedValueToUse);
}
```
we modify `resolvedValueToUse` and don't pass original `resolvedValue` downstream.
Unlike here we have this:
```
if (resolvedValue instanceof String[]) {
	for (Object object : (String[]) resolvedValue) {
		resolveAsString(object, result);
	}
}
if (resolvedValue instanceof String) {
	result.add((String) resolvedValue);
}
```
so, we continue to pass original value downstream the method.
A bit different logic.

**Comment by garyrussell** [comment]
Right; that's why I don't understand how it is different; we used to have the `resolvedValueToUse` variable here but it was still wrong, and you removed it during polishing at some point. Looks like I simply screwed up, and we didn't notice when you did your polishing.

**Comment by artembilan** [comment]
Yeah... Maybe.
I guess I just decided that we don't need that extra recursion over artificial `Arrays.asList((String[]) resolvedValue)`.
Anyway good to have a report on the matter and quick fix.

I don't see reason to seek a guilty man. It's not so difficult to fix and the workaround is like:
```
@KafkaListener(topics = "#{T(org.springframework.util.CollectionUtils).arrayToList('${kafka.inbound.topics}'.split(','))}")
```
`Arrays.asList()` doesn't work in SpEL, because it knows nothing about the type of our value and treats it as a single vararg argument.

**Comment by umekong** [comment]
Thanks guys.

**Comment by girish-1234** [comment]
I have list of topics configured in a string 
for e.g. String topicnames = "topic1,topic2,topic3"

The requirement is we also need to specify partition alongwith the topic names 
@KafkaListener(id = "abc", topicPartitions =
        { @TopicPartition(topic = "topic1", partitions = "0"),
          @TopicPartition(topic = "topic2", partitions = "0"),
          @TopicPartition(topic = "topic3", partitions = "0"),
        })
Can someone help me to understand if it is possible to use Spring Expression Language here?

**Comment by artembilan** [comment]
@girish-1234 ,

first of all you comment on a closed ticket. So, it might be just missed. Because of closed status.
Another concern that your comment is a question. 
That would be much better to do that on community forums. 
We use GitHub for bug reports and feature requests.

Now I'll try to answer to your question with what the best I can suggest. 

What you want with dynamic topics and their partitions assignments isn't possible with annotations.
Now you have 
```
{ @topicpartition(topic = "topic1", partitions = "0"),
@topicpartition(topic = "topic2", partitions = "0"),
@topicpartition(topic = "topic3", partitions = "0"),
```
but you will have one more topic, you would need to add one more `@TopicPartition` annotation to this `@KafkaListener` definition.

You can achieve that only via functional programming, not restricted static abilities with annotations.

The `@KafkaListener` has option like:
```
	/**
	 * The bean name of the {@link org.springframework.kafka.config.KafkaListenerContainerFactory}
	 * to use to create the message listener container responsible to serve this endpoint.
	 * <p>If not specified, the default container factory is used, if any.
	 * @return the container factory bean name.
	 */
	String containerFactory() default "";
```
What I'd suggest is some customization for the `ConcurrentKafkaListenerContainerFactory` and modification to the `createContainerInstance()` method with the population of desired `TopicPartitionInitialOffset` set according to your config.

**Comment by artembilan** [comment]
Sorry, your question is somehow unclear and doesn't follow public communication principles.
You should be specific as much as as possible.

I'm just following your previous comment:

> (more topics added in conf file)

No, you can't do that because you can't add more `@TopicPartition` in the compiled already application.

Would be much better to use Java code for this kind of tasks. 
We have just spent with you plenty of time discussing the dead end of the annotations abilities.
Believe me, it's pointless to pull from annotations what they don't support.

**Comment by girish-1234** [comment]
I have created topicpartitions and now i want to set it as part of ConcurrentKafkaListenerContainerFactory.
However if i want to override createContainerInstance of  ConcurrentKafkaListenerContainerFactory,
it expects KafkaListenerEndpoint where i need to specify all methods implementation.
It would be great help if you could point me to a sample code to customize ConcurrentKafkaListenerContainerFactory. 

TopicPartitionInitialOffset topicPartitions1 = new TopicPartitionInitialOffset("topic1", 0);
        TopicPartitionInitialOffset topicPartitions2 = new TopicPartitionInitialOffset("topic2", 0);
        TopicPartitionInitialOffset topicPartitions3 = new TopicPartitionInitialOffset("topic3", 0);
        
 ContainerProperties containerProperties = new 
 ContainerProperties(topicPartitions1,topicPartitions2,topicPartitions3);
 KafkaMessageListenerContainer<String, String> container = createContainer(containerProperties);

**Comment by artembilan** [comment]
Please, consider to investigate how to write readable comments on GitHub.
Right now it is hard to parse a mix of code and normal words in your comment.

What I suggested is something like this:
````java
@Override
protected ConcurrentMessageListenerContainer createContainerInstance(KafkaListenerEndpoint endpoint) {
	TopicPartitionInitialOffset topicPartitions1 = new TopicPartitionInitialOffset("topic1", 0);
	TopicPartitionInitialOffset topicPartitions2 = new TopicPartitionInitialOffset("topic2", 0);
	TopicPartitionInitialOffset topicPartitions3 = new TopicPartitionInitialOffset("topic3", 0);
	((AbstractKafkaListenerEndpoint) endpoint).setTopicPartitions(topicPartitions1, topicPartitions2, topicPartitions3);
							
	return super.createContainerInstance(endpoint);
}
````

**Comment by girish-1234** [comment]
Thank you so much.
